pkg/agent/controller/packetcapture/bpf/parser.go

```diff
@@ -0,0 +1,154 @@
+		inst = append(inst, bpf.JumpIf{Cond: bpf.JumpEqual, Val: addrVal, SkipTrue: 0, SkipFalse: size - uint8(len(inst)) - 2})
+	}
+
+	srcPort := matchPacket.SourcePort
+	dstPort := matchPacket.DestinationPort
+	if srcPort > 0 || dstPort > 0 {
+		skipTrue := size - uint8(len(inst)) - 3
+		inst = append(inst, loadIPv4HeaderOffset(skipTrue)...)
+		if srcPort > 0 {
+			inst = append(inst, loadIPv4SourcePort)
+			inst = append(inst, bpf.JumpIf{Cond: bpf.JumpEqual, Val: uint32(srcPort), SkipTrue: 0, SkipFalse: size - uint8(len(inst)) - 2})
+		}
+		if dstPort > 0 {
+			inst = append(inst, loadIPv4DestinationPort)
+			inst = append(inst, bpf.JumpIf{Cond: bpf.JumpEqual, Val: uint32(dstPort), SkipTrue: 0, SkipFalse: size - uint8(len(inst)) - 2})
+		}
+
+	}
+
+	// return
+	inst = append(inst, returnKeep)
+	inst = append(inst, returnDrop)
+
+	return inst
+
+}
+
+// we need to figure out how long the instruction list will be first. will be used in the instructions'jump case.
+// For example, If you provide all the filter supported by `PacketCapture`, it will ends with the following BPF filter string:
+// 'ip proto 6 and src host 127.0.0.1 and dst host 127.0.0.1 and src port 123 and dst port 124'
+// And using `tcpdump -i <device> '<filter>' -d` will generate the following BPF instructions:
+// (000) ldh      [12]                                     # Load 2B at 12 (Ethertype)
+// (001) jeq      #0x800           jt 2	jf 16              # Ethertype: If IPv4, goto #2, else #16
+// (002) ldb      [23]                                     # Load 1B at 23 (IPv4 Protocol)
+// (003) jeq      #0x6             jt 4	jf 16              # IPv4 Protocol: If TCP, goto #4, #16
+// (004) ld       [26]                                     # Load 4B at 26 (source address)
+// (005) jeq      #0x7f000001      jt 6	jf 16              # If bytes match(127.0.0.1), goto #6, else #16
+// (006) ld       [30]                                     # Load 4B at 30 (dest address)
+// (007) jeq      #0x7f000001      jt 8	jf 16              # If bytes match(127.0.0.1), goto #8, else #16
+// (008) ldh      [20]                                     # Load 2B at 20 (13b Fragment Offset)
+// (009) jset     #0x1fff          jt 16	jf 10      # Use 0x1fff as a mask for fragment offset; If fragment offset != 0, #10, else #16
+// (010) ldxb     4*([14]&0xf)                             # x = IP header length
+// (011) ldh      [x + 14]                                 # Load 2B at x+14 (TCP Source Port)
+// (012) jeq      #0x7b            jt 13	jf 16      # TCP Source Port: If 123, goto #13, else #16
+// (013) ldh      [x + 16]                                 # Load 2B at x+16 (TCP dst port)
+// (014) jeq      #0x7c            jt 15	jf 16      # TCP dst port: If 123, goto $15, else #16
+// (015) ret      #262144                                  # MATCH
+// (016) ret      #0                                       # NOMATCH
+
+func calInstructionsSize(packet *crdv1alpha1.Packet) int {
```

Comment:
is it `calInstructionsSize` or `callInstructionsSize`?

Reply:
`cal`. i want to calculate the length of the final instructions.